src/zero_copy/input.rs

```diff
@@ -35,13 +35,8 @@ impl Input for str {
     }
 
     fn next(&self, offset: Self::Offset) -> (Self::Offset, Option<Self::Token>) {
-        if offset < self.len() {
-            // TODO: Can we `unwrap_unchecked` here?
-            let c = unsafe { self.get_unchecked(offset..).chars().next().unwrap() };
-            (offset + c.len_utf8(), Some(c))
-        } else {
-            (offset, None)
-        }
+        let chr = self.chars().skip(offset).next();
+        (offset + chr.map_or(0, char::len_utf8), chr)
```

Comment:
I'd prefer if this change were not made because this is extremely hot code. Even single branches will have a substantial impact on parser performance (ideally. `unwrap_unchecked` would be used too: `str` is guaranteed to always be valid UTF-8, so provided we can guarantee that `offset` is valid, this should be fine).

Reply:
With your comment in mind, I went ahead and ran both variations through Godbolt , and came to the conclusion that my suggestion had about 3x the amount of branches that yours did :D So, good call there. I was curious though what a hybrid variation would look like, and ended up doing this:

```rust
fn next(&self, offset: Self::Offset) -> (Self::Offset, Option<Self::Token>) {
    let chr = unsafe { self.get_unchecked(offset..).chars().next() };
    (offset + chr.map_or(0, char::len_utf8), chr)
}
```

The output of Godbolt can be found [here](https://rust.godbolt.org/z/3oxh9YhTs), and shows that this variation has one less branch than the original implementation. I am curious on if you would have any benchmarks that could be used to determine which of the two implementations proves faster!

Reply:
That seems like it might be better in that case! The `json` benchmark is currently the canonical way to benchmark chumsky (although it would be nice to have others in the future given that it doesn't include cases like backtracking).

Reply:
I just realized why it would have one less branch, and I think we both missed something when we thought we could merge it. I believe we have to revert the changes here.

Assuming the called passes in an offset that is too large, its no longer going to return none, but instead read memory that it shouldn't. I spaced that `get_unchecked` couldn't be used without a check, so that's my bad.